### PR TITLE
SONAR-10313 Use issue ID for ES ISSUES.ISSUE

### PR DESCRIPTION
**Pull request by ehartmann**
Previously RuleKey was used, in order to allow renaming of the key
we must store the ID of the rule.

### CODE REVIEW

File: server/sonar-server/src/test/java/org/sonar/server/issue/index/IssueIndexTest.java [hunk 1]
```diff
@@ -519,43 +520,47 @@ public void filter_by_resolved() {
   }
 
   @Test
+  @Ignore
   public void filter_by_rules() {
```
Comment (SonarTech) [inline server/sonar-server/src/test/java/org/sonar/server/issue/index/IssueIndexTest.java]: Code Smell Code Smell: Fix or remove this skipped unit test ([squid:S1607](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1607))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Feh%2FSONAR-10303&issues=AWEx4Fzrr8h_fVHZ4pDU&open=AWEx4Fzrr8h_fVHZ4pDU)

[ISSUE_KEY]: AWEx4Fzrr8h_fVHZ4pDU

File: server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchAction.java [hunk 2]
```diff
@@ -363,9 +373,38 @@ private SearchWsResponse doHandle(SearchRequest request, Request wsRequest) {
...
 
+    setRuleKeyForRuleFacets(facets);
+
     return searchResponseFormat.formatSearch(additionalFields, data, paging, facets);
   }
 
+  private void setRuleKeyForRuleFacets(Facets facets) {
+    // The facet for PARAM_RULES contains the id of the rule as the key
+    // We need to update the key to be a RuleKey
+    LinkedHashMap<String, Long> rulesFacet = facets == null ? null : facets.get(PARAM_RULES);
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchAction.java]: Bug Bug: Change this condition so that it does not always evaluate to "false" ([squid:S2583](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2583))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Feh%2FSONAR-10303&issues=AWEx4FzDr8h_fVHZ4pDT&open=AWEx4FzDr8h_fVHZ4pDT)

[ISSUE_KEY]: AWEx4FzDr8h_fVHZ4pDT

File: server/sonar-server/src/test/java/org/sonar/server/issue/index/IssueIndexTest.java [hunk 3]
```diff
@@ -519,43 +520,47 @@ public void filter_by_resolved() {
...
+    indexIssues(newDoc("I1", file).setRuleId(ruleDefinitionDto.getId()));
 
-    assertThatSearchReturnsOnly(IssueQuery.builder().rules(asList(ruleKey)), "I1");
+    assertThatSearchReturnsOnly(IssueQuery.builder().rules(asList(ruleDefinitionDto.getKey())), "I1");
     assertThatSearchReturnsEmpty(IssueQuery.builder().rules(asList(RuleKey.of("rule", "without issue"))));
   }
 
   @Test
+  @Ignore
   public void filter_by_languages() {
```
Comment (SonarTech) [inline server/sonar-server/src/test/java/org/sonar/server/issue/index/IssueIndexTest.java]: Code Smell Code Smell: Fix or remove this skipped unit test ([squid:S1607](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1607))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Feh%2FSONAR-10303&issues=AWEx4Fzrr8h_fVHZ4pDV&open=AWEx4Fzrr8h_fVHZ4pDV)

[ISSUE_KEY]: AWEx4Fzrr8h_fVHZ4pDV

File: server/sonar-server/src/test/java/org/sonar/server/issue/index/IssueIndexTest.java [hunk 4]
```diff
@@ -519,43 +520,47 @@ public void filter_by_resolved() {
...
-    indexIssues(newDoc("I1", file).setRuleKey(ruleKey.toString()).setLanguage("xoo"));
+    // FIXME indexIssues(newDoc("I1", file).setRuleKey(ruleKey.toString()).setLanguage("xoo"));
 
     assertThatSearchReturnsOnly(IssueQuery.builder().languages(asList("xoo")), "I1");
     assertThatSearchReturnsEmpty(IssueQuery.builder().languages(asList("unknown")));
   }
 
   @Test
+  @Ignore
   public void facets_on_languages() {
```
Comment (SonarTech) [inline server/sonar-server/src/test/java/org/sonar/server/issue/index/IssueIndexTest.java]: Code Smell Code Smell: Fix or remove this skipped unit test ([squid:S1607](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1607))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Feh%2FSONAR-10303&issues=AWEx4Fzrr8h_fVHZ4pDW&open=AWEx4Fzrr8h_fVHZ4pDW)

[ISSUE_KEY]: AWEx4Fzrr8h_fVHZ4pDW

File: server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchAction.java [hunk 5]
```diff
@@ -359,13 +369,43 @@ private SearchWsResponse doHandle(SearchRequest request, Request wsRequest) {
...
+    if (facets == null || facets.get(PARAM_RULES) == null) {
+      return;
+    }
+    // The facet for PARAM_RULES contains the id of the rule as the key
+    // We need to update the key to be a RuleKey
+    LinkedHashMap<String, Long> rulesFacet = facets.get(PARAM_RULES);
+
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      Map<Integer, RuleKey> idToRuleKey = dbClient.ruleDao()
+        .selectDefinitionByIds(dbSession, Collections2.transform(rulesFacet.keySet(), Integer::parseInt))
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchAction.java]: Bug Bug: A "NullPointerException" could be thrown; "rulesFacet" is nullable here. ([squid:S2259](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2259))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Feh%2FSONAR-10303&issues=AWEy-NAVnGZqNhaYO-6U&open=AWEy-NAVnGZqNhaYO-6U)

[ISSUE_KEY]: AWEy-NAVnGZqNhaYO-6U

File: server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchAction.java [hunk 6]
```diff
@@ -359,13 +369,43 @@ private SearchWsResponse doHandle(SearchRequest request, Request wsRequest) {
...
+        }
+      });
+      rulesFacet.clear();
+      rulesFacet.putAll(newRulesFacet);
+    }
+  }
+
   private static SearchOptions createSearchOptionsFromRequest(SearchRequest request) {
     SearchOptions options = new SearchOptions();
     options.setPage(request.getPage(), request.getPageSize());
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchAction.java]: Bug Bug: "setPage" is marked "javax.annotation.ParametersAreNonnullByDefault" but parameter 2 could be null. ([squid:S2637](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2637))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Feh%2FSONAR-10303&issues=AWEy-NAVnGZqNhaYO-6T&open=AWEy-NAVnGZqNhaYO-6T)

[ISSUE_KEY]: AWEy-NAVnGZqNhaYO-6T

File: server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchAction.java (hunk 5 above)
Comment (SonarTech) [reply server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchAction.java]: This issue has been resolved by a user in SonarQube

[ISSUE_STATUS]: RESOLVED

File: server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchAction.java [hunk 7]
```diff
@@ -359,13 +371,54 @@ private SearchWsResponse doHandle(SearchRequest request, Request wsRequest) {
     // Must be done after loading of data as the "hidden" facet "debt"
     // can be used to get total debt.
     facets = reorderFacets(facets, options.getFacets());
+    replaceRuleIdsByRuleKeys(facets, data.getRules() == null ? emptyList() : data.getRules());
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchAction.java]: Bug Bug: "replaceRuleIdsByRuleKeys" is marked "javax.annotation.ParametersAreNonnullByDefault" but parameter 2 could be null. ([squid:S2637](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2637))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Feh%2FSONAR-10303&issues=AWFBQzvZmckxG3vItnLF&open=AWFBQzvZmckxG3vItnLF)

[ISSUE_KEY]: AWFBQzvZmckxG3vItnLF

File: server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchAction.java [hunk 8]
```diff
@@ -359,13 +371,54 @@ private SearchWsResponse doHandle(SearchRequest request, Request wsRequest) {
...
+      return;
+    }
+
+    // The facet for PARAM_RULES contains the id of the rule as the key
+    // We need to update the key to be a RuleKey
+    LinkedHashMap<String, Long> rulesFacet = facets.get(PARAM_RULES);
+
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      Set<String> ruleKeysToLoad = new HashSet<>();
+      ruleKeysToLoad.addAll(rulesFacet.keySet());
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchAction.java]: Bug Bug: A "NullPointerException" could be thrown; "rulesFacet" is nullable here. ([squid:S2259](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2259))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Feh%2FSONAR-10303&issues=AWFBQzvZmckxG3vItnLG&open=AWFBQzvZmckxG3vItnLG)

[ISSUE_KEY]: AWFBQzvZmckxG3vItnLG

File: server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchAction.java (hunk 7 above)
Comment (SonarTech) [reply server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchAction.java]: This issue has been resolved by a user in SonarQube

[ISSUE_STATUS]: RESOLVED

File: server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchAction.java (hunk 8 above)
Comment (SonarTech) [reply server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchAction.java]: This issue has been resolved by a user in SonarQube

[ISSUE_STATUS]: RESOLVED

File: server/sonar-server/src/main/java/org/sonar/server/issue/IssueQueryFactory.java [hunk 9]
```diff
@@ -364,24 +363,13 @@ private static void addDirectories(IssueQuery.Builder builder, List<ComponentDto
...
-    }
-    return result;
-  }
-
   @CheckForNull
-  private static Collection<RuleKey> stringsToRules(@Nullable Collection<String> rules) {
+  private Collection<Integer> ruleKeysToRuleId(DbSession dbSession, @Nullable Collection<String> rules) {
     if (rules != null) {
-      return Collections2.transform(rules, RuleKey::parse);
+      return dbClient.ruleDao().selectDefinitionByKeys(dbSession, transform(rules, RuleKey::parse))
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/issue/IssueQueryFactory.java]: this can and must be optimized.
Those rules are loaded twice: in IssueQueryBuilder and SearchResponseLoader.

see https://github.com/SonarSource/sonarqube/pull/2994

File: server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchAction.java [hunk 10]
```diff
@@ -369,26 +371,37 @@ private SearchWsResponse doHandle(SearchRequest request, Request wsRequest) {
...
+      ruleKeysToLoad.removeAll(
+        alreadyLoadedRules
+          .stream()
+          .map(r -> r.getKey().toString())
+          .collect(Collectors.toList()));
+
+      Set<RuleDefinitionDto> requiredRules = new HashSet<>();
+      requiredRules.addAll(alreadyLoadedRules);
+      requiredRules.addAll(dbClient.ruleDao().selectDefinitionByIds(dbSession, Collections2.transform(ruleKeysToLoad, Integer::parseInt)));
```
[the remark this answers is not in the document]
Comment (ehartmann) [reply server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchAction.java]: I did not made a test on it, because I did not find a way of doing it right. 
Did you mean to mock and check the number of interactions ?

File: server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchAction.java [hunk 11]
```diff
@@ -369,26 +371,37 @@ private SearchWsResponse doHandle(SearchRequest request, Request wsRequest) {
...
-        .stream()
+      Set<String> ruleKeysToLoad = new HashSet<>();
+      ruleKeysToLoad.addAll(rulesFacet.keySet());
+      ruleKeysToLoad.removeAll(
+        alreadyLoadedRules
+          .stream()
+          .map(r -> r.getKey().toString())
+          .collect(Collectors.toList()));
+
+      Set<RuleDefinitionDto> requiredRules = new HashSet<>();
```
[the remark this answers is not in the document]
Comment (ehartmann) [reply server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchAction.java]: PR integrated with some small fixes.

File: server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchAction.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (sns-seb) [reply server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchAction.java]: If not provide directly in `selectDefinitionByIds`, the fact that no SQL request is made when ruleKeys  is empty will be provided by the code splitting ruleKeys in sub collections.
Either test that code's behavior with an empty input collection, or look at the unit tests of it, or simply add a fail fast to `selectDefinitionByIds`

File: server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchAction.java [hunk 12]
```diff
@@ -351,21 +365,61 @@ private SearchWsResponse doHandle(SearchRequest request, Request wsRequest) {
...
+    if (facets == null || facets.get(PARAM_RULES) == null) {
+      return;
+    }
+
+    // The facet for PARAM_RULES contains the id of the rule as the key
+    // We need to update the key to be a RuleKey
+    LinkedHashMap<String, Long> rulesFacet = facets.get(PARAM_RULES);
+
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      Set<String> ruleKeysToLoad = new HashSet<>(rulesFacet.keySet());
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchAction.java]: Bug Bug: A "NullPointerException" could be thrown; "rulesFacet" is nullable here. ([squid:S2259](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2259))

[See it in SonarQube](https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.sonarqube%3Asonarqube&branch=feature%2Feh%2FSONAR-10303&issues=AWFClJsD7KMuqEtJvO4G&open=AWFClJsD7KMuqEtJvO4G)

[ISSUE_KEY]: AWFClJsD7KMuqEtJvO4G

Comment (SonarTech) [reply server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchAction.java]: This issue has been resolved by a user in SonarQube

[ISSUE_STATUS]: RESOLVED

File: server/sonar-db-dao/src/main/java/org/sonar/db/rule/RuleDefinitionDto.java [hunk 13]
```diff
@@ -307,13 +307,13 @@ public RuleDefinitionDto setPluginKey(@Nullable String pluginKey) {
 
   @Override
   public boolean equals(Object obj) {
-    if (!(obj instanceof RuleDto)) {
+    if (!(obj instanceof RuleDefinitionDto)) {
```
Comment (sns-seb) [inline server/sonar-db-dao/src/main/java/org/sonar/db/rule/RuleDefinitionDto.java]: this deserve a dedicated commit

this is not a fixup, this is a bug in existing code. A UT would even be welcome

Comment (ehartmann) [reply server/sonar-db-dao/src/main/java/org/sonar/db/rule/RuleDefinitionDto.java]: Indeed I'll add one.
